Core: Fix Java Time DateFormatter printers

spinscale opened the pull request:
A bug in the test suite prevented to properly check that all date
formatters printed the date the same way like joda time does.

This fixes the test and thus also a fair share of formats, that
now use the strict parser for printing.

Also, now all date parsers are using UTC as their default time zone.

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented on server/src/main/java/org/elasticsearch/common/time/CompoundDateTimeFormatter.java:
```diff
@@ -70,4 +76,68 @@ public CompoundDateTimeFormatter withZone(ZoneId zoneId) {
+        return new CompoundDateTimeFormatterBuilder();
+    }
+
+    static final class CompoundDateTimeFormatterBuilder {
```
Instead of creating another class, could the functionality here just be a couple private static methods on DateFormatters? If I understand these timezone formats correctly, the only reason we have any of this complexity is because of `+HHmm`, which is a non standard format. The other two would be handled by `appendZoneOrOffsetId()`? I think it would be good to keep this local to DateFormatters, so that adding more cases like this does not leak outside of that class.

rjernst commented on server/src/main/java/org/elasticsearch/common/time/DateFormatters.java:
```diff
@@ -332,7 +268,9 @@
-        .append(OPTIONAL_TIME_ZONE_FORMATTER)
+        .optionalStart().appendZoneId().optionalEnd()
+        .optionalStart().appendOffset("+HHmm", "Z").optionalEnd()
+        .optionalStart().appendOffset("+HH:mm", "Z").optionalEnd()
```
Can you use `appendZoneOrOffsetId()` above, which includes this?

spinscale commented:
I removed the methods from the `CompoundDateFormatter`, used `appendZoneOrOffsetId()` and also did not create more static methods, but created the date time formatters manually (the methods did not save that much).

I also removed making all the formatters UTC by default, as there seem to be differences between java8 and 10 (java10 works as expected, java8 fails the tests in CI and local).

spinscale commented:
I also added a small comment for each of the compound date time formatters now
